client/purplecaffeine/widget.py

```diff
@@ -0,0 +1,337 @@
+        detail_view (DomWidget): widget to display the details of a trial
+        pagination_view (DomWidget): widget to display the pagination buttons
+    """
+
+    def __init__(self, storage: Optional[BaseStorage] = None):
+        """Widget class:
+        Attributes:
+            storage (BaseStorage): storage where the trials are going to be saved
+        """
+        self.storage = storage or LocalStorage("./trials")
+        self.limit = 10
+        self.offset = 0
+        self.current_page_trials: List[Trial] = self.storage.list(
+            limit=self.limit, offset=self.offset
+        )
+        self.next_page_trials: List[Trial] = self.storage.list(
+            limit=self.offset + self.limit + self.limit, offset=self.offset + self.limit
+        )
+        self.selected_trial: Optional[Trial] = (
+            self.current_page_trials[0] if len(self.current_page_trials) > 0 else None
+        )
+        self.search_value = ""
+
+        self.list_view = widgets.Output()
+        with self.list_view:
+            display(self.render_trails_list())
+
+        self.detail_view = widgets.Output()
+        with self.detail_view:
+            display(self.render_trial())
+
+        self.pagination_view = widgets.Output()
+        with self.pagination_view:
+            display(self.render_pagination())
+
+    def display_message(self, required_message):
+        """
+        Returns a user-friendly message for displaying
+        Returns:
+            empty_message (HTMLWidget): the widget with the info
+        """
+        empty_message = widgets.HTML(
+            f"<h1 style='text-align: center;'> <br><br><br>{required_message}</h1>"
+            '<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/'
+            'dist/css/bootstrap.min.css" '
+            'integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" '
+            'crossorigin="anonymous">'
+        )
+        empty_message.layout = Layout(height="300px")
+        return empty_message
+
+    def load_detail(self, trial_button):
+        """
+        Load the details of a trial
+            Args:
+            trial_button: a trial button that was clicked
+        """
+        trial_id = trial_button.tooltip
+        trial = self.storage.get(trial_id)
+        if isinstance(trial, Trial):
+            self.selected_trial = trial
+            with self.detail_view:
+                clear_output()
+                display(self.render_trial())
+        else:
+            self.display_message("Something went wrong while loading your trials ):")
+
+    def render_trails_list(self):
+        """
+        Render the list of trail buttons.
+        We render one button per each trial
+        """
+        buttons = []
+        for trial in self.current_page_trials:
+            button = widgets.Button(
+                description=f"{trial.name} | {trial.uuid}"[:30],
```

Comment:
Maybe do not show the uuid in the title, that's not very user friendly

Reply:
no, it's ok. Sometimes this information is needed on high level